### ISSUE TITLE
LOADING Redis is loading the dataset in memory

### ISSUE THREAD

**Issue by turbo124** [predis/predis#1144]
**Describe the bug**
When using redis-sentinel, if the slave network cannot maintain throughput to stay within the client-output-buffer limits, the slave will disconnect with the following

```
Predis\Response\ServerException
LOADING Redis is loading the dataset in memory
```

This issue has been in the library since 1.x
https://github.com/predis/predis/pull/475

The above PR provided a graceful fallback to the master in these scenarios where the slaves were temporarily unable to return read requests.  When evaluating 2.x of this library, we are still able to reproduce these scenarios resulting in exceptions rather than graceful fallback to master

**Expected behavior**
There are a couple of options for this scenario.

- Provide a configuration option to only serve requests to the master. Load balancing is not the primary goal of using a Redis-sentinel.
- Patch using a similar approach where in the catch, if the exception matches, we serve the request from the master

```php

    private function retryCommandOnFailure(CommandInterface $command, $method)
    {
        $retries = 0;

        while ($retries <= $this->retryLimit) {
            try {
                $response = $this->getConnectionByCommand($command)->$method($command);
                break;
            } catch(\Predis\Response\ServerException $exception){

                if($exception->getMessage() == 'LOADING Redis is loading the dataset in memory'){
                    $response = $this->getMaster()->$method($command);
                }

            } catch (CommunicationException $exception) {

                
                $this->wipeServerList();
                $exception->getConnection()->disconnect();

                if ($retries === $this->retryLimit) {
                    throw $exception;
                }

                usleep($this->retryWait * 1000);

                ++$retries;
            }
        }

        return $response;
    }

```

Would a solution like this be considered? Or is there another approach recommended so that we can get this issue resolved and merged into 2.x ?

**Versions (please complete the following information):**
- Predis: [e.g. 1.x 2.x]
- PHP [e.g. 8.1.14]
- Redis Server [e.g. 7.*]
- OS [e.g. Ubuntu 20.04]

**Comment by tillkruss** [comment]
What do the Redis Sentinel specs say about failing commands during `LOADING`? I always treated it as the server being down.

Ping @chayim.

**Comment by tillkruss** [comment]
I'd say in that case let's handle in better on our end 👍

@vladvildanov Any thoughts?

**Comment by vladvildanov** [comment]
@tillkruss Yeah, I like an idea to handle this on our side, also agreed that it should configurable option. The whole idea of replication is to be able to retrieve data from multiple sync sources, so it's probably should be enabled by default to switch on master on connection issues

**Comment by chayim** [comment]
The loading state indicates that things are backed, due to the window size on the master. The -LOADING response to ping is valid in this case.

IMHO the client shouldn't handle it, but should provide the option to the user. In some cases you want to know this is happening, and handle in on your application side. Other applications may not care - basically it depends what you're doing.

But - a retry handle could be a great answer here. This would allow the user to handle the case with their own function implementation.

**Comment by turbo124** [comment]
@chayim 

Pushing this into userland would be problematic, especially those using frameworks (for us Laravel) as we'd then be reliant on a solution being accepted/provided there.

At the very least the client should provide an option to define the read/write nodes, and not force load balancing in a redis-sentinel configuration.

**Comment by mta59066** [comment]
So am I understanding this correctly; 

Predis SentinelReplication if it gets a connection error it will retry the command but if it gets a LOADING state, it will throw the exception, while predis [MasterSlaveReplication](https://github.com/predis/predis/blob/v2.x/src/Connection/Replication/MasterSlaveReplication.php#L484) will switch to next replica or finally to the master before throwing an exception and the proposed change here is to do the same in SentinelReplication just like it is done in MasterSlaveReplication, but that was not agreed somehow? Is that right?

**Comment by turbo124** [comment]
@mta59066 

Currently the default operation of this library will load balance all read requests across the master and all read replicas. There are *some* checks for replica health, however p_sync state is kinda *fuzzy* and even thou the replica will return a LOADING state, the library will still send a request to that replica....

The PR here: #1168 allows you to force all read/writes to the master which is the work around.

**Comment by mta59066** [comment]
@turbo124 what you proposed initially is exactly how [MasterSlaveReplication](https://github.com/predis/predis/blob/v2.x/src/Connection/Replication/MasterSlaveReplication.php#L484) works and I don't see why sacrificing on load balancing is needed to be able to handle the LOADING case on sentinel. By using SentinelReplication userland has already given that control to the client which retires the command (default 20 times). Perhaps we should just copy the retryCommandOnFailure method from MasterSlaveReplication and have them behave in the same way. Is there something I am missing between SentinelReplication and MasterSlaveReplication that should make them behave differently?

**Comment by turbo124** [comment]
Current checks are only for sdown, not psync. *that* is the crux of the issue, and tbh, I am surprised upstream would even consider a node in partial resync as not sdown'd.

Another solution would be to add an additional check for psync status.

**Comment by mta59066** [comment]
above mr is merged, this issue can be closed
